modules/nlu/src/backend/application/training-queue.ts

```diff
@@ -0,0 +1,300 @@
+import * as sdk from 'botpress/sdk'
+
+import _ from 'lodash'
+
+import ms from 'ms'
+import { ITrainingRepository } from './training-repo'
+import { ITrainingService } from './training-service'
+import {
+  TrainingId,
+  TrainerService,
+  TrainingListener,
+  TrainingState,
+  TrainingSession,
+  I,
+  LockedTrainingSession
+} from './typings'
+
+export interface TrainingQueueOptions {
+  maxTraining: number
+}
+
+const DEFAULT_OPTIONS: TrainingQueueOptions = {
+  maxTraining: 2
+}
+
+const DEFAULT_STATUS: TrainingState = { status: 'idle', progress: 0 }
+
+export type ITrainingQueue = I<TrainingQueue>
+
+const debug = DEBUG('nlu').sub('lifecycle')
+
+const TRAINING_LIFE_TIME = ms('5m')
+
+class LocalTrainingContainer {
+  private _localTrainings: TrainingId[] = []
+
+  public length() {
+    return this._localTrainings.length
+  }
+
+  public async startLocalTraining(id: TrainingId) {
+    this._localTrainings.push(id)
+  }
+
+  public async rmLocalTraining(id: TrainingId) {
+    const isTraining = (t: TrainingId) => this._areEqual(t, id)
+    this._localTrainings = this._localTrainings.filter(_.negate(isTraining))
+  }
+
+  private _areEqual(id1: TrainingId, id2: TrainingId) {
+    return id1.botId === id2.botId && id1.language === id2.language
+  }
+}
+
+export class TrainingQueue implements TrainingQueue {
```

Comment:
I don't see the TrainingQueue interface anywhere, how can this compile ? 

ITrainingQueue is built by interfacing this Class ...

Reply:
leftover... good catch.

As of right now it implements itself... Strangly, this is valid Typescript